Misfit preprocessor

markusdregi opened the pull request:
Resolves: https://github.com/equinor/semeio/issues/68

**Misson of the PR**
- Establish the user interface of the observation correlation tooling that is to pushed out in the company as described in #68.
- Add a bigger test base to ensure that results are as expected. In particular I've mocked parts of the `MeasuredData` structure from `ERT`, which allows me to quickly evaluate thousands of forward models without depending on libres.

**Remaining work for this PR**
- Find a better name for the workflow. The idea of `AUTO_CORRELATE_OBSERVATIONS` was that it would be semi-automatic. However, this is a very overloaded term in statistics, so we should call it something else. Perhaps `SCALE_OBSERVATION_CLUSTERS`?

**Remaining work**
- Write documentation. I've started filling out the schema. Will try to use configsuite tooling afterwards to generate a first iteration.
- Unite the schema of spearman correlation in this workflow and the spearman job.

jondequinor commented:
Meeeeh.. :man_shrugging: I don't know man…

Anyway, be aware of the quite major refactoring going on in https://github.com/equinor/semeio/pull/66

markusdregi commented:
I did not really get the meh?

markusdregi commented:
Regarding the refactoring, I’m keeping a close look on it and if I manage to get anything done it will be on the config side until you merge your work to avoid conflicts

jondequinor commented on README.md:
```diff
@@ -1,3 +1,5 @@
 
 Jobs and workflow jobs used in [ERT](https://github.com/equinor/ert).
+
+....
```
This was the _meh_.

markusdregi replied:
How come? It is the perfect dummy commit 🥳

jondequinor replied:
Please use ellipsis …

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+    "correlation",
+    "cosine",
+    "dice",
+    _EUCLIDEAN,
```
it's quite many metrics, are all the metrics really necessary to be exposed? If yes I would perhaps suggest this (in order not to define name of each single metric):
```python
from enum import Enum

class Metrics(Enum):
         euclidean=0
         cityblock=1

# then you can access name
Metrics.euclidean.name
```

markusdregi replied:
In general I aimed for a very thin abstraction on top of `fcluster` and `linkage` such that the users can be guided to the `scipy` documentation and expect to find the same arguments exposed in the configuration. I could of course remove some of the metrics, but all that would save me was to define the strings... The values is propagated in the same manner no matter if we support only three of them or all.

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+@cs.validator_msg("t must be an integer if a max cluster criteria is choosen")
+def _t_is_int_if_maxclust(spearman_config):
+    if spearman_config[_CRITERION] in (_MAXCLUST, _MAXCLUST_MONOCRIT):
+        return isinstance(spearman_config[_T], int)
```
what's "t"? Can we replace it with something more descriptive?

xjules replied:
_CLUSTER_THRESHOLD = "t" ?

markusdregi replied:
Same argument as above. The `scipy` docs refer to the argument as `t`. I'll write up the explanations and then there will be text explaining the value in the docs. Again, we can change the name and/or behaviour, but then we are not purely exposing the `scipy` parameters anymore. Which is also fine 🤷

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+def _bounds_validator(
+    lower=None, lower_inclusive=True, upper=None, upper_inclusive=True
+):
+    msg = "Value must be within the interval {}, {}".format(
```
Very nice! It took me a while to parse this :)

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+}
+
+
+_CLUSTERING_SCHEMA = {
```
Is the separation of `_CLUSTERING_SCHEMA` and `_SPEARMAN_CORRELATION_SCHEMA` due to readability or something else? Also why the name `_SPEARMAN_CORRELATION_SCHEMA` when it deals only with `fcluster` in scipy?

markusdregi replied:
The separation is purely for readability and specifically to avoid 10 indents in a dictionary. I named it `_SPEARMAN_CORRELATION_SCHEMA` because it represent all the parameters that can be configured for our clustering algorithm based on spearman correlation. Happy to change the name if you have a suggestion :)

xjules replied:
Lazy to bring up a good name besides `OBSERVATION_CLUSTERING` so we leave at that :)

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+}
+
+
+_SCALING_SCHEMA = {
```
I reckon this schema is work in progress.

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+
+
+# TODO: Should we rename this one?
+class AutoCorrelateConfig(cs.ConfigSuite):
```
Hm and why not `AutoScalingConfig`?

markusdregi replied:
I think we should change the name away from `Auto` and then consistently change these names.

xjules commented on semeio/jobs/scripts/auto_correlate_observations.py:
```diff
@@ -0,0 +1,51 @@
+
+class AutoCorrelateObservationsJob(ErtScript):
+    def run(self, *args):
+        config_record = _fetch_config_record(args)
```
Why we need an extra function `_fetch_config_record` as this only loads the file?

markusdregi replied:
Because I'm designing for replacement and abstraction, not for reuse :) And since I'm later not expecting `args` to point to files, but to contain a stream of data, I've separated that part of the code from the core logic 🤷

xjules commented on semeio/jobs/auto_correlate_observations/job.py:
```diff
@@ -0,0 +1,35 @@
+from semeio.jobs.spearman_correlation_job.job import spearman_job
+
+
+def run(auto_correlation_config, measured_data, **records):
```
What's records?

markusdregi replied:
Nothing besides allowing the general (and non-existent) framework to pass more records then the two we actually need. But I think it is a good idea to remove it until it actually exists 👍

markusdregi commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+                _CRITERION: {
+                    cs.MetaKeys.Type: cs.types.String,
+                    cs.MetaKeys.ElementValidators: (
+                        _one_of(
```
Well, I really need all of the members to be strings (as I want to write `inconsistent` nor `0` in the configuration file). Also, I have been thinking about implementing a `one_of` type in `configsuite` and wanted to make a first test of the idea here.

xjules replied:
The string is implicitly provided on the fly with `.name` property:
```python
print(Criterion.inconsistent.name)
# "inconsistent"

# validation for string exist
string_exist = "inconsistent" in Criterion.__members__
```

markusdregi replied:
I would prefer to not place them in an enum as a container the api in configsuite will only support containers (at least in the first iteration). It is quite a bit more flexible than the limited amounts of types that enums support :)

But I agree that name spacing is good. So what I've done is to collect them into a tuple at least...

markusdregi commented on semeio/jobs/spearman_correlation_job/job.py:
```diff
@@ -4,7 +4,14 @@
-def spearman_job(measured_data, threshold):
+def spearman_job(
+    measured_data,
+    t,
```
I don't mind. I just changed it consistently to cohere with the `scipy` docs to make the job a really thin wrapper 🤷

We can deviate from this in the code, or also in the UI...

markusdregi commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+_METHOD = "method"
+_SPEARMAN_CORRELATION = "spearman_correlation"
+_FCLUSTER = "fcluster"
+_T = "t"
```
Sure 👍

markusdregi commented:
> I think this looks good so far. But how will you prevent duplicating the SC and COS config schemas, and all it's validation/transformation logic?
>
> One way to solve this would be to let ACO absorb both SC and COS—make them submodules of the former. The user would not lose the ability to run either SC or COS in isolation.

I'm not certain @jondequinor 🤔 I just wanted to make a first shoot at the UI first and then tackle the rest later. One option is to have ACO supersede both SC and COS, but then the UI will have to be expanded to also allow for manually specifying the groups that are to be scaled. Another option would be to have `SC` and `COS` expose their schemas and then plugg them into the schema of ACO.

xjules commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,315 @@
+class _ObservationContext:
+    def __init__(self, observation_keys):
+        Context = collections.namedtuple("Context", ("observation_keys",))
+        self._context = Context(set(observation_keys))
```
Why this needs to be a set? Are we expecting an observation key to appear more than once?

markusdregi replied:
Nope, it just seemed like an appropriate data structure if iteration and containment testing is what the user of the data structure is supposed to use it for.

xjules commented on semeio/jobs/spearman_correlation_job/job.py:
```diff
@@ -87,6 +96,6 @@ def _calculate_correlation_matrix(data):
     return data.rank().corr(method="pearson")
```
Can't find where this data comes from?

xjules commented on tests/jobs/auto_correlate_observations/test_integration.py:
```diff
@@ -0,0 +1,159 @@
+@pytest.mark.usefixtures("setup_tmpdir")
+def test_auto_correlate_skip_clusters_yielding_empty_data_matrixes(monkeypatch):
+    def raising_scaling_job(data):
+        if data == {"CALCULATE_KEYS": {"keys": [{"index": [88, 89], "key": "FOPR"}]}}:
```
This one would benefit from a comment or two, since it's hard to see what's happening.

xjules commented on tests/jobs/auto_correlate_observations/unit/test_auto_correlate_observations.py:
```diff
@@ -0,0 +1,239 @@
+from semeio.jobs import auto_correlate_observations
+
+
+class MockedMeasuredData(object):
```
Could you make a comment on the data we mock here?

markusdregi replied:
We mock the `MeasuredData` object in ERT. Would you like me to put a reference to it?

xjules commented on tests/jobs/auto_correlate_observations/unit/test_auto_correlate_observations.py:
```diff
@@ -0,0 +1,239 @@
+        pass
+
+
+def assertHomogenClusters(config):
```
Why camelCase? Also why this is called homogenCluster?

markusdregi replied:
camelCase due to old habit of following unittest syntax for assert functions... `homogenCluster` because it asserts that all members of each cluster comes from the same polynomial.

ManInFez commented on semeio/jobs/auto_correlate_observations/config.py:
```diff
@@ -0,0 +1,398 @@
+                    cs.MetaKeys.Type: cs.types.String,
+                    cs.MetaKeys.Description: (
+                        "The criterion to use in forming flat clusters. "
+                        "Defaults to {}."
```
Name or not named parameters for formatting?

markusdregi replied:
I don't have any strong preferences. Do you?

ManInFez commented on semeio/jobs/auto_correlate_observations/job.py:
```diff
@@ -0,0 +1,31 @@
+        )
+
+    config = config.snapshot
+    if config.clustering.method == "spearman_correlation":
```
should fetch name from config namespace maybe?

ManInFez commented on semeio/jobs/scripts/auto_correlate_observations.py:
```diff
@@ -0,0 +1,75 @@
+
+    scale_conf = config.snapshot.scaling
+    return {
+        "threshold": scale_conf.threshold,
```
fetch strings from config namespace

markusdregi replied:
This string actually comes from the COS job and that schema consists of magic strings. So there is no possibility of fetching this without doing some work in the schema. See https://github.com/equinor/semeio/blob/e6a958997703d8bcb2cef3c57de05f8034c87681/semeio/jobs/correlated_observations_scaling/job_config.py#L120

Persist it in an issue instead?

ManInFez replied:
👍

ManInFez commented on semeio/jobs/spearman_correlation_job/job.py:
```diff
@@ -4,7 +4,14 @@
+def spearman_job(
+    measured_data,
+    threshold,
+    criterion="inconsistent",
```
Perhaps shared information about default values should be accessible for all jobs?

markusdregi replied:
Yes, as @jondequinor also suggested, they should probably share schema and default values. I can either resolve that here or we can leave that for a follow up, smaler PR 🤷 Let me know what you prefer...

ManInFez replied:
make a follow-up issue, and we are golden

markusdregi commented:
Thanks 👍 I'll change the name before merging it...

markusdregi commented:
@ManInFez: I've rebased the PR and removed all traces of `Auto Correlate Observations` at the benefit of `Misfit Preprocessor`. I've grep'ed all commits for `auto`, verified that the tests pass on all commits and the line diff count on GitHub is preserved (the last is slightly lucky as I did not have to run black due to too large change in line length). I've also searched for auto here on GitHub in the diff.

I think it is good to go, but I'll await for your agreement before I merge.

ManInFez commented:
I concur!
